liquibase/README.md

```diff
@@ -0,0 +1,340 @@
+
+### Change sets in (hypothetic) OpenMRS 4.7.x
+Looking forward to a (hypothetic) version 4.7.x of OpenMRS, the respective change sets are:
+
+* `org/openmrs/liquibase/snapshots/schema-only/liquibase-schema-only-4.7.x.xml` defines the OpenMRS schema. This file is a **snapshot** generated from OpenMRS 4.7.x.
+
+* `org/openmrs/liquibase/snapshots/core-data/liquibase-core-data-4.7.x.xml` defines core data. Again, this file is a **snapshot** generated from OpenMRS 4.7.x.
+
+* `org/openmrs/liquibase/updates/liquibase-update-to-latest-4.8.x.xml` contains database changes introduced by OpenMRS 4.8.x
+
+### Further Liquibase files 
+* `liquibase-update-to-latest.xml` is used by integration tests and includes references to multiple `org/openmrs/liquibase/updates/liquibase-update-to-latest-a.b.x.xml` files.
+  
+* `liquibase-empty-changelog.xml` is used as a default Liquibase file by the org.openmrs.util.DatabaseUpdater class.
+
+## When to generate Liquibase shapshots
+Liquibase snapshots need to be created...
+
+1. when a **new minor or major version** of OpenMRS is created (such as 2.3.x or 3.0.x), new snapshot files need to be 
+generated for the **previous** versions. The new snapshot files are added to
+ 
+    * `org/openmrs/liquibase/snapshots/schema-only` and 
+    * `org/openmrs/liquibase/snapshots/core-data` 
+
+   in the OpenMRS **master branch**. Do not forget to include the version number in the change log filenames.
+  
+   Liquibase updates introduced with the new 
+version are added to
+
+    * `org/openmrs/liquibase/updates`
+   
+   The examples for the (hypothetic) OpemMRS version 4.7.x further above illustrates the different version numbers to use for the new change log files.
+
+2. when a **database change is added to an existing minor or major version**, the snapshot files of later versions 
+need to be updated so that they include the change.
+
+## Generating and applying Liquibase snapshots
+The pom file of the openmrs-liquibase module contains a template for generating Liquibase snapshots from an existing database and applying snapshots to an OpenMRS database.
+
+### How to generate Liquibase snapshots
+#### Step 1 - Drop your local OpenMRS schema
+E.g. by running the script `openmrs-core/liquibase/scripts/drop_openmrs_schema.sql`:
+
+	mysql -u root -p openmrs < openmrs-core/liquibase/scripts/drop_openmrs_schema.sql
+	
+Take care **not** to run this script on a production database.
+
+#### Step 2 - Build and initialise OpenMRS
+	cd <some root folder>/openmrs-core
+	mvn clean install
+	cd webapp
+	rm -r ~/.OpenMRS
+	mvn jetty:run
+
+Open [http://localhost:8080/openmrs/initialsetup](http://localhost:8080/openmrs/initialsetup) and choose the following options:
+
+* **simple installation** in step 2 of the installation wizard
+
+* **not to add demo data** in step 3 of the installation wizard
+
+#### Step 3 - Create a snapshot of the OpenMRS schema
+	cd <some root folder>/openmrs-core/liquibase
```

Comment:
Do you remember how long this command took to finish on your computer? For me, it takes around 3hours. Is that the expected time frame?

Reply:
No, definitely not for hours, I need to re-run this when removing the old jar files from the pom.xml but it was maybe some 10 minutes.

Reply:
What kind of machine are you using pls?

Reply:
MacBook Pro (Retina, 15-inch, Mid 2015)
Processor 2.8 GHz Intel Core i7
Memory 16 GB 1600 MHz DDR3

Reply:
```
~/Projects/OpenMRS/code10/openmrs-core/liquibase(TRUNK-4830-N) $ mvn -DoutputChangelogfile=liquibase-schema-only-SNAPSHOT.xml -Dusername=root -Dpassword=Admin123 liquibase:generateChangeLog
[INFO] Scanning for projects...
[INFO] 
...
[INFO] changeSets count: 881
[INFO] ../api/src/main/resources/liquibase-schema-only-SNAPSHOT.xml does not exist, creating and adding 881 changesets.
[INFO] Output written to Change Log file, ../api/src/main/resources/liquibase-schema-only-SNAPSHOT.xml
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:35 min
[INFO] Finished at: 2020-05-12T17:44:02+02:00
[INFO] ------------------------------------------------------------------------
~/Projects/OpenMRS/code10/openmrs-core/liquibase(TRUNK-4830-N) $ mvn -DdiffTypes=data -DoutputChangelogfile=liquibase-core-data-SNAPSHOT.xml -Dusername=root -Dpassword=Admin123 liquibase:generateChangeLog
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------< org.openmrs.liquibase:openmrs-liquibase >---------------
[INFO] Building openmrs-liquibase 2.4.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- liquibase-maven-plugin:3.8.9:generateChangeLog (default-cli) @ openmrs-liquibase ---
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] 
[INFO] Liquibase Community 3.8.9 by Datical
[INFO] Starting Liquibase at Tue, 12 May 2020 17:46:53 CEST (version 3.8.9 #73 built at Mon Apr 06 09:32:58 UTC 2020)
[INFO] Executing on Database: jdbc:mysql://127.0.0.1:3306/openmrs
[INFO] Generating Change Log from database root@localhost @ jdbc:mysql://127.0.0.1:3306/openmrs (Default Schema: openmrs)
[INFO] changeSets count: 27
[INFO] ../api/src/main/resources/liquibase-core-data-SNAPSHOT.xml does not exist, creating and adding 27 changesets.
[INFO] Output written to Change Log file, ../api/src/main/resources/liquibase-core-data-SNAPSHOT.xml
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.047 s
[INFO] Finished at: 2020-05-12T17:46:56+02:00
[INFO] ------------------------------------------------------------------------
~/Projects/OpenMRS/code10/openmrs-core/liquibase(TRUNK-4830-N) $ 

```

Reply:
01:30 minutes for schema-only and 3 seconds for core-data.

Reply:
And it worked well w/o the legacy libraries, so I can remove them from the module's pom file.

@dkayiwa please hold your fire before you re-run the generation of snapshots for a few minutes and use the next commit.

Reply:
The update is out, without the old modify column jar (which ticks the second box on @ibacher 's list)

Reply:
Re hardware I am using an 2018 Macbook Pro, the 2015 model should be perfectly ok.

Reply:
A few colleagues of mine had tested the generation of snapshots and they would have let me know if it had taken ages for them...